qubes/tests/vm/qubesvm.py

```diff
@@ -1968,8 +1971,90 @@ def test_624_qdb_guivm_invalid_keyboard_layout(self, mock_qubesdb,
-            guivm.features['keyboard-layout'] = 'fr123++'
+        with self.assertRaises(qubes.exc.QubesPropertyValueError):
+            guivm.keyboard_layout = 'fr123++'
+
+        with self.assertRaises(qubes.exc.QubesPropertyValueError):
+            guivm.keyboard_layout = 'fr+???+'
+
+        with self.assertRaises(qubes.exc.QubesPropertyValueError):
+            guivm.keyboard_layout = 'fr++variant?'
+
+        with self.assertRaises(qubes.exc.QubesPropertyValueError):
+            guivm.keyboard_layout = 'fr'
+
+    @unittest.mock.patch('qubes.utils.get_timezone')
+    @unittest.mock.patch('qubes.utils.urandom')
+    @unittest.mock.patch('qubes.vm.qubesvm.QubesVM.untrusted_qdb')
+    def test_625_qdb_keyboard_layout_change(self, mock_qubesdb, mock_urandom,
+            mock_timezone):
+        mock_urandom.return_value = b'A' * 64
+        mock_timezone.return_value = 'UTC'
+        template = self.get_vm(
+            cls=qubes.vm.templatevm.TemplateVM, name='template')
+        template.netvm = None
+        guivm = self.get_vm(cls=qubes.vm.appvm.AppVM, template=template,
+            name='sys-gui', qid=2, provides_network=False)
+        vm = self.get_vm(cls=qubes.vm.appvm.AppVM, template=template,
+            name='appvm', qid=3)
+        vm.netvm = None
+        vm.guivm = guivm
+        guivm.keyboard_layout = 'fr++'
+        guivm.is_running = lambda: True
+        vm.events_enabled = True
+        test_qubesdb = TestQubesDB()
+        mock_qubesdb.write.side_effect = test_qubesdb.write
+        mock_qubesdb.rm.side_effect = test_qubesdb.rm
+        vm.create_qdb_entries()
+        self.maxDiff = None
+
+        expected = {
+            '/name': 'test-inst-appvm',
+            '/type': 'AppVM',
+            '/default-user': 'user',
+            '/keyboard-layout': 'fr++',
+            '/qubes-vm-type': 'AppVM',
+            '/qubes-gui-domain-xid': '{}'.format(guivm.xid),
+            '/qubes-debug-mode': '0',
+            '/qubes-base-template': 'test-inst-template',
+            '/qubes-timezone': 'UTC',
+            '/qubes-random-seed': base64.b64encode(b'A' * 64),
+            '/qubes-vm-persistence': 'rw-only',
+            '/qubes-vm-updateable': 'False',
+            '/qubes-block-devices': '',
+            '/qubes-usb-devices': '',
+            '/qubes-iptables': 'reload',
+            '/qubes-iptables-error': '',
+            '/qubes-iptables-header': unittest.mock.ANY,
+            '/qubes-service/qubes-update-check': '0',
+            '/qubes-service/meminfo-writer': '1',
+            '/connected-ips': '',
+            '/connected-ips6': '',
+        }
+
+        with self.subTest('default'):
+            vm.create_qdb_entries()
+            self.assertEqual(test_qubesdb.data, expected)
+
+        with self.subTest('value_change'):
+            vm.keyboard_layout = 'de++'
+            expected['/keyboard-layout'] = 'de++'
+            vm.create_qdb_entries()
```

Comment:
Event handler should make this call unnecessary (and its's worth checking if that's really the case).